File: fabric-sdk-js/src/identity/fif.js
// FIF parse/unlock/serialize. Wire format spec:
//   docs/specs/fabric-spec-001-v1.0.md §"Fabric Identity File"
//
// Interop note: AAD for the envelope's AEAD is the on-wire envelope-header
// bytes (4-byte big-endian length + UTF-8 JSON). The Go SDK does the same;
// fabric-sdk-go/identity/fif.go documents the binding.

import {
  seal,
  open,
  randomNonce,
  randomSalt,
  deriveKeyArgon2id,
  DEFAULT_ARGON2ID_PARAMS,
  SALT_SIZE,
  NONCE_SIZE,
  TAG_SIZE,
} from '../crypto.js';
import { bytesToBase64, base64ToBytes } from '../util/base64.js';

export const FIF_FORMAT = 'fif/1.0';
export const INNER_FIF_FORMAT = 'fif-inner/1.0';

/** v1.0 envelope types: only "passphrase-only" is accepted; rest are reserved. */
export const ENVELOPE_PASSPHRASE_ONLY = 'passphrase-only';
export const RESERVED_ENVELOPE_TYPES = Object.freeze([
  'shamir-shares',
  'device-quorum',
  'social-recovery',
]);

/**
 * FIFError carries a protocol-aligned code so callers can branch on
 * fif_format / fif_auth / fif_envelope_unsupported / identity_locked.
 */
export class FIFError extends Error {
  constructor(code, message, cause) {
    super(message);
    this.name = 'FIFError';
    this.code = code;
    if (cause) this.cause = cause;
  }
}

const enc = new TextEncoder();
const dec = new TextDecoder('utf-8', { fatal: true });

// Inner-FIF binary fields that travel as base64 strings on the wire. Helpers
// below convert them to Uint8Array on parse and back to base64 on serialize.
const SCALAR_BINARY_PATHS = [
  ['root_keypair', 'public_key'],
  ['root_keypair', 'private_key'],
];
const ARRAY_BINARY_PATHS = [
  ['device_subkeys', 'public_key'],
  ['device_subkeys', 'private_key'],
  ['agent_identities', 'public_key'],
  ['agent_identities', 'private_key'],
  ['transports', 'public_key'],
  ['grants_held', 'macaroon'],
];

function reviveInnerBinary(inner) {
  for (const [parent, field] of SCALAR_BINARY_PATHS) {
    const v = inner?.[parent]?.[field];
    if (typeof v === 'string') inner[parent][field] = base64ToBytes(v);
  }
  for (const [arr, field] of ARRAY_BINARY_PATHS) {
    if (Array.isArray(inner?.[arr])) {
      for (const item of inner[arr]) {
        if (typeof item?.[field] === 'string') item[field] = base64ToBytes(item[field]);
      }
    }
  }
  return inner;
}

function dehydrateInnerBinary(inner) {
  // Return a deep-cloned representation where Uint8Array fields are base64
  // strings, so JSON.stringify produces a wire-compatible body.
  const out = JSON.parse(
    JSON.stringify(inner, (_, v) => {
      if (v instanceof Uint8Array) return bytesToBase64(v);
      return v;
    }),
  );
  return out;
}

/** Build an InnerFIF skeleton with collections initialized to empties. */
export function newInnerFIF(principal, rootKeypair) {
  return {
    format: INNER_FIF_FORMAT,
    principal,
    root_keypair: rootKeypair,
    device_subkeys: [],
    agent_identities: [],
    transports: [],
    grants_held: [],
    bridge_credentials: [],
    recent_state_cache: {
      vault_heads: {},
      history_heads: {},
    },
  };
}

/** Build a parsed FIF in memory with a fresh passphrase-only envelope. */
export async function newFIF(passphrase, inner) {
  if (!inner) throw new FIFError('fif_format', 'inner is required');
  const salt = randomSalt();
  const nonce = randomNonce();
  const params = DEFAULT_ARGON2ID_PARAMS;
  const key = await deriveKeyArgon2id(passphrase, salt, params);
  const header = {
    format: FIF_FORMAT,
    envelope_type: ENVELOPE_PASSPHRASE_ONLY,
    envelope_params: {
      kdf: 'argon2id',
      kdf_params: {
        m_cost: params.memory,
        t_cost: params.time,
        parallelism: params.parallelism,
      },
      salt: bytesToBase64(salt),
      nonce: bytesToBase64(nonce),
    },
  };
  return new FIF({ header, key, inner, salt, nonce });
}

/**
 * ParseFIF reads the envelope header and retains the encrypted body. Call
 * unlock() to decrypt. Throws FIFError with code "fif_envelope_unsupported"
 * if the envelope_type is reserved or unknown.
 *
 * @param {Uint8Array} bytes
 * @returns {FIF}
 */
export function parseFIF(bytes) {
  if (bytes.length < 4) {
    throw new FIFError('fif_format', 'too short for length prefix');
  }
  const view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength);
  const headerLen = view.getUint32(0, false /* big-endian */);
  if (headerLen === 0 || headerLen > 1 << 20) {
    throw new FIFError('fif_format', `header length out of range: ${headerLen}`);
  }
  if (bytes.length < 4 + headerLen + TAG_SIZE) {
    throw new FIFError('fif_format', 'body shorter than AEAD tag');
  }
  const headerJSON = bytes.subarray(4, 4 + headerLen);
  const body = bytes.subarray(4 + headerLen);
  let header;
  try {
    header = JSON.parse(dec.decode(headerJSON));
  } catch (e) {
    throw new FIFError('fif_format', 'header JSON parse failed', e);
  }
  if (header.format !== FIF_FORMAT) {
    throw new FIFError('fif_format', `unknown format ${JSON.stringify(header.format)}`);
  }
  if (RESERVED_ENVELOPE_TYPES.includes(header.envelope_type)) {
    throw new FIFError(
      'fif_envelope_unsupported',
      `envelope_type ${JSON.stringify(header.envelope_type)} is reserved for v1.x`,
    );
  }
  if (header.envelope_type !== ENVELOPE_PASSPHRASE_ONLY) {
    throw new FIFError(
      'fif_envelope_unsupported',
      `envelope_type ${JSON.stringify(header.envelope_type)} is not supported`,
    );
  }
  if (header.envelope_params?.kdf !== 'argon2id') {
    throw new FIFError('fif_format', `unsupported kdf ${header.envelope_params?.kdf}`);
  }
  const salt = base64ToBytes(header.envelope_params.salt);
  const nonce = base64ToBytes(header.envelope_params.nonce);
  if (salt.length !== SALT_SIZE) {
    throw new FIFError('fif_format', `salt length ${salt.length}, want ${SALT_SIZE}`);
  }
  if (nonce.length !== NONCE_SIZE) {
    throw new FIFError('fif_format', `nonce length ${nonce.length}, want ${NONCE_SIZE}`);
  }
  // Retain the raw header bytes as AAD (the bytes the issuer used, verbatim).
  const headerBytes = bytes.subarray(0, 4 + headerLen);
  return new FIF({ header, headerBytes, body, salt, nonce });
}

/** Internal: assembles a 4-byte big-endian length prefix + bytes. */
function withLengthPrefix(jsonBytes) {
  const out = new Uint8Array(4 + jsonBytes.length);
  new DataView(out.buffer).setUint32(0, jsonBytes.length, false);
  out.set(jsonBytes, 4);
  return out;
}

export class FIF {
  /**
   * @param {{header: object, key?: Uint8Array, inner?: object,
   *          headerBytes?: Uint8Array, body?: Uint8Array,
   *          salt: Uint8Array, nonce: Uint8Array}} parts
   */
  constructor(parts) {
    this.header = parts.header;
    this.key = parts.key ?? null;
    this.inner = parts.inner ?? null;
    this.headerBytes = parts.headerBytes ?? null;
    this.body = parts.body ?? null;
    this.salt = parts.salt;
    this.nonce = parts.nonce;
  }

  get envelopeType() {
    return this.header.envelope_type;
  }

  isUnlocked() {
    return this.inner !== null && this.key !== null;
  }

  async unlock(passphrase) {
    if (this.isUnlocked()) return;
    if (!this.body || !this.headerBytes) {
      throw new FIFError('fif_format', 'cannot unlock — FIF was not parsed from bytes');
    }
    const p = this.header.envelope_params.kdf_params;
    const key = await deriveKeyArgon2id(passphrase, this.salt, {
      time: p.t_cost,
      memory: p.m_cost,
      parallelism: p.parallelism,
      hashLength: 32,
    });
    let plaintext;
    try {
      plaintext = open(key, this.nonce, this.body, this.headerBytes);
    } catch (e) {
      throw new FIFError('fif_auth', 'FIF authentication failed', e);
    }
    let inner;
    try {
      inner = JSON.parse(dec.decode(plaintext));
    } catch (e) {
      throw new FIFError('fif_format', 'parse inner JSON', e);
    }
    if (inner.format !== INNER_FIF_FORMAT) {
      throw new FIFError('fif_format', `unknown inner format ${inner.format}`);
    }
    reviveInnerBinary(inner);
    this.key = key;
    this.inner = inner;
  }

  serialize() {
    if (!this.isUnlocked()) {
      throw new FIFError('identity_locked', 'FIF is locked');
    }
    const innerWire = dehydrateInnerBinary(this.inner);
    const innerJSON = enc.encode(JSON.stringify(innerWire));
    const headerJSON = enc.encode(JSON.stringify(this.header));
    const headerBytes = withLengthPrefix(headerJSON);
    const ciphertext = seal(this.key, this.nonce, innerJSON, headerBytes);
    const out = new Uint8Array(headerBytes.length + ciphertext.length);
    out.set(headerBytes, 0);
    out.set(ciphertext, headerBytes.length);
    this.headerBytes = headerBytes;
    this.body = ciphertext;
    return out;
  }

  lock() {
    if (this.key) this.key.fill(0);
    this.key = null;
    this.inner = null;
  }
}

File: fabric-sdk-go/identity/fif.go
package identity

import (
	"bytes"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"

	"github.com/NakliTechie/private-mesh/fabric-sdk-go/crypto"
)

// FIFFormat is the format string carried by the FIF envelope header.
const FIFFormat = "fif/1.0"

// EnvelopeType is the user-recovery envelope wrapping the inner FIF.
type EnvelopeType string

const (
	EnvelopePassphraseOnly EnvelopeType = "passphrase-only"
	// Reserved for v1.x. v1.0 refuses these with ErrFIFEnvelopeUnsupported.
	EnvelopeShamirShares   EnvelopeType = "shamir-shares"
	EnvelopeDeviceQuorum   EnvelopeType = "device-quorum"
	EnvelopeSocialRecovery EnvelopeType = "social-recovery"
)

// reservedEnvelopeTypes are valid envelope_type values that v1.0 implementations
// must explicitly refuse with ErrFIFEnvelopeUnsupported (handoff §"Forward-compatibility hooks").
var reservedEnvelopeTypes = map[EnvelopeType]struct{}{
	EnvelopeShamirShares:   {},
	EnvelopeDeviceQuorum:   {},
	EnvelopeSocialRecovery: {},
}

// envelopeHeader is the JSON shape of the envelope's leading header.
type envelopeHeader struct {
	Format         string                  `json:"format"`
	EnvelopeType   EnvelopeType            `json:"envelope_type"`
	EnvelopeParams passphraseOnlyEnvParams `json:"envelope_params"`
}

// passphraseOnlyEnvParams matches envelope_params for envelope_type="passphrase-only".
type passphraseOnlyEnvParams struct {
	KDF       string         `json:"kdf"`
	KDFParams argon2idParams `json:"kdf_params"`
	Salt      []byte         `json:"salt"`
	Nonce     []byte         `json:"nonce"`
}

// argon2idParams is the JSON shape of envelope_params.kdf_params.
type argon2idParams struct {
	MCost       uint32 `json:"m_cost"`
	TCost       uint32 `json:"t_cost"`
	Parallelism uint8  `json:"parallelism"`
}

func (p argon2idParams) toCrypto() crypto.Argon2idParams {
	return crypto.Argon2idParams{
		Time:    p.TCost,
		Memory:  p.MCost,
		Threads: p.Parallelism,
		KeyLen:  crypto.KeySize,
	}
}

func argon2idParamsFrom(p crypto.Argon2idParams) argon2idParams {
	return argon2idParams{
		MCost:       p.Memory,
		TCost:       p.Time,
		Parallelism: p.Threads,
	}
}

// FIF is a parsed Fabric Identity File. After Unlock or NewFIF, Inner is populated.
type FIF struct {
	// header holds the parsed envelope header.
	header envelopeHeader
	// headerBytes is the raw on-wire envelope-header (4-byte length + JSON).
	// It is used as the AEAD AAD so a tampered header invalidates the MAC.
	headerBytes []byte
	// body is the encrypted body || tag (still encrypted when freshly parsed).
	body []byte
	// key is the derived envelope key, cached after Unlock so Serialize can
	// re-encrypt without re-running the KDF. Zeroed by Lock.
	key []byte
	// Inner is the decrypted body. Nil until Unlock or NewFIF.
	Inner *InnerFIF
}

// EnvelopeType returns the envelope type of this FIF (without unlocking).
func (f *FIF) EnvelopeType() EnvelopeType { return f.header.EnvelopeType }

// IsUnlocked reports whether Inner is populated and the envelope key is cached.
func (f *FIF) IsUnlocked() bool { return f.Inner != nil && len(f.key) == crypto.KeySize }

// NewFIF builds an unlocked FIF in memory using a fresh passphrase-only envelope.
// Call Serialize to write it out.
func NewFIF(passphrase string, inner *InnerFIF) (*FIF, error) {
	if inner == nil {
		return nil, fmt.Errorf("identity.NewFIF: inner is nil")
	}
	salt, err := crypto.RandomSalt()
	if err != nil {
		return nil, err
	}
	nonce, err := crypto.RandomNonce()
	if err != nil {
		return nil, err
	}
	kdfParams := crypto.DefaultArgon2idParams()
	key := crypto.DeriveKeyArgon2id(passphrase, salt, kdfParams)

	return &FIF{
		header: envelopeHeader{
			Format:       FIFFormat,
			EnvelopeType: EnvelopePassphraseOnly,
			EnvelopeParams: passphraseOnlyEnvParams{
				KDF:       "argon2id",
				KDFParams: argon2idParamsFrom(kdfParams),
				Salt:      salt,
				Nonce:     nonce,
			},
		},
		key:   key,
		Inner: inner,
	}, nil
}

// ParseFIF reads the envelope header and retains the encrypted body. Call Unlock
// to decrypt. ParseFIF refuses unknown or reserved envelope_type values with
// ErrFIFEnvelopeUnsupported.
func ParseFIF(r io.Reader) (*FIF, error) {
	var lenBuf [4]byte
	if _, err := io.ReadFull(r, lenBuf[:]); err != nil {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("read header length: %w", err))
	}
	headerLen := binary.BigEndian.Uint32(lenBuf[:])
	if headerLen == 0 || headerLen > 1<<20 {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("header length out of range: %d", headerLen))
	}
	headerJSON := make([]byte, headerLen)
	if _, err := io.ReadFull(r, headerJSON); err != nil {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("read header: %w", err))
	}
	body, err := io.ReadAll(r)
	if err != nil {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("read body: %w", err))
	}
	if len(body) < crypto.TagSize {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("body shorter than AEAD tag"))
	}

	var hdr envelopeHeader
	dec := json.NewDecoder(bytes.NewReader(headerJSON))
	dec.DisallowUnknownFields()
	if err := dec.Decode(&hdr); err != nil {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("parse header JSON: %w", err))
	}
	if hdr.Format != FIFFormat {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("unknown format %q", hdr.Format))
	}
	if _, reserved := reservedEnvelopeTypes[hdr.EnvelopeType]; reserved {
		return nil, codedError(ErrFIFEnvelopeUnsupported, fmt.Errorf("envelope_type %q is reserved for v1.x", hdr.EnvelopeType))
	}
	if hdr.EnvelopeType != EnvelopePassphraseOnly {
		return nil, codedError(ErrFIFEnvelopeUnsupported, fmt.Errorf("envelope_type %q is not supported", hdr.EnvelopeType))
	}
	if hdr.EnvelopeParams.KDF != "argon2id" {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("unsupported kdf %q", hdr.EnvelopeParams.KDF))
	}
	if len(hdr.EnvelopeParams.Salt) != crypto.SaltSize {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("salt length %d, want %d", len(hdr.EnvelopeParams.Salt), crypto.SaltSize))
	}
	if len(hdr.EnvelopeParams.Nonce) != crypto.NonceSize {
		return nil, codedError(ErrFIFFormat, fmt.Errorf("nonce length %d, want %d", len(hdr.EnvelopeParams.Nonce), crypto.NonceSize))
	}

	// Rebuild the on-wire header bytes for AEAD AAD. We use the bytes we read
	// rather than re-marshalling, so any whitespace-level differences across
	// implementations are preserved verbatim and the MAC verifies.
	headerBytes := make([]byte, 4+len(headerJSON))
	binary.BigEndian.PutUint32(headerBytes[:4], headerLen)
	copy(headerBytes[4:], headerJSON)

	return &FIF{
		header:      hdr,
		headerBytes: headerBytes,
		body:        body,
	}, nil
}

// Unlock derives the envelope key from passphrase, decrypts the body, and
// populates f.Inner. The key is cached so Serialize can re-encrypt without
// asking for the passphrase again. Returns ErrFIFAuth on wrong passphrase.
func (f *FIF) Unlock(passphrase string) error {
	if f.Inner != nil && f.IsUnlocked() {
		return nil
	}
	params := f.header.EnvelopeParams.KDFParams.toCrypto()
	key := crypto.DeriveKeyArgon2id(passphrase, f.header.EnvelopeParams.Salt, params)

	plaintext, err := crypto.Open(key, f.header.EnvelopeParams.Nonce, f.body, f.headerBytes)
	if err != nil {
		return codedError(ErrFIFAuth, err)
	}
	var inner InnerFIF
	if err := json.Unmarshal(plaintext, &inner); err != nil {
		return codedError(ErrFIFFormat, fmt.Errorf("parse inner JSON: %w", err))
	}
	if inner.Format != InnerFIFFormat {
		return codedError(ErrFIFFormat, fmt.Errorf("unknown inner format %q", inner.Format))
	}
	f.key = key
	f.Inner = &inner
	return nil
}

// Serialize encrypts the current Inner under the cached envelope key and writes
// the full FIF (4-byte length + header JSON + ciphertext+tag) to w. Returns
// ErrIdentityLocked if Unlock or NewFIF have not been called.
func (f *FIF) Serialize(w io.Writer) error {
	if !f.IsUnlocked() {
		return ErrIdentityLocked
	}
	innerJSON, err := json.Marshal(f.Inner)
	if err != nil {
		return fmt.Errorf("marshal inner: %w", err)
	}
	headerJSON, err := json.Marshal(f.header)
	if err != nil {
		return fmt.Errorf("marshal header: %w", err)
	}
	headerBytes := make([]byte, 4+len(headerJSON))
	binary.BigEndian.PutUint32(headerBytes[:4], uint32(len(headerJSON)))
	copy(headerBytes[4:], headerJSON)

	ciphertext, err := crypto.Seal(f.key, f.header.EnvelopeParams.Nonce, innerJSON, headerBytes)
	if err != nil {
		return err
	}
	if _, err := w.Write(headerBytes); err != nil {
		return err
	}
	if _, err := w.Write(ciphertext); err != nil {
		return err
	}
	// Persist the canonical header bytes so subsequent Unlock(serialized) on the
	// same in-memory FIF would use the same AAD if anyone re-reads.
	f.headerBytes = headerBytes
	f.body = ciphertext
	return nil
}

// Lock zeroes the cached envelope key and drops the in-memory Inner. The FIF
// must be Unlock()ed again before any further crypto operations.
func (f *FIF) Lock() {
	for i := range f.key {
		f.key[i] = 0
	}
	f.key = nil
	f.Inner = nil
}

var _ error = (*Error)(nil)
